core/shared/src/main/scala/zio/stm/TQueue.scala

```diff
@@ -18,77 +18,83 @@ package zio.stm
-  def offer(a: A): STM[Nothing, Unit] =
-    for {
-      q <- ref.get
-      _ <- STM.check(q.length < capacity)
-      _ <- ref.update(_.enqueue(a))
-    } yield ()
-
-  // TODO: Scala doesn't allow Iterable???
-  @silent("enqueueAll")
-  def offerAll(as: List[A]): STM[Nothing, Unit] =
-    ref.update(_.enqueue(as)).unit
-
-  def poll: STM[Nothing, Option[A]] = takeUpTo(1).map(_.headOption)
-
-  def size: STM[Nothing, Int] = ref.get.map(_.length)
 
-  def take: STM[Nothing, A] =
-    ref.get.flatMap { q =>
-      q.dequeueOption match {
-        case Some((a, as)) =>
-          ref.set(as) *> STM.succeed(a)
-        case _ => STM.retry
-      }
-    }
-
-  def takeAll: STM[Nothing, List[A]] =
-    ref.modify(q => (q.toList, ScalaQueue.empty[A]))
+  /**
+   * Checks if the queue is empty.
+   */
+  def isEmpty: STM[Nothing, Boolean] =
+    ref.get.map(_.isEmpty)
 
-  def takeUpTo(max: Int): STM[Nothing, List[A]] =
-    ref.get
-      .map(_.splitAt(max))
-      .flatMap(split => ref.set(split._2) *> STM.succeed(split._1))
-      .map(_.toList)
+  /**
+   * Checks if the queue is at capacity.
+   */
+  def isFull: STM[Nothing, Boolean] =
+    ref.get.map(_.size == capacity)
 
   /**
-   * View the next element in the queue without removing it. retries if the queue is empty.
+   * Views the last element inserted into the queue, retrying if the queue is
+   * empty.
    */
-  def peek: STM[Nothing, A] =
+  def last: STM[Nothing, A] =
     ref.get.flatMap(
-      _.headOption match {
+      _.lastOption match {
         case Some(a) => STM.succeed(a)
         case None    => STM.retry
       }
     )
 
   /**
-   * View the last element inserted into the queue. retries if the queue is empty
+   * Offers the specified value to the queue, retrying if the queue is at
+   * capacity.
    */
-  def last: STM[Nothing, A] =
+  def offer(a: A): STM[Nothing, Unit] =
+    ref.get.flatMap { q =>
+      if (q.length < capacity) ref.update(_.enqueue(a)).unit
+      else STM.retry
+    }
+
+  /**
+   * Offers each of the elements in the specified collection to the queue,
+   * retrying if there is not capacity in the queue for all of the elements in
+   * the collection.
+   */
+  def offerAll(as: Iterable[A]): STM[Nothing, Unit] =
+    ref.get.flatMap { q =>
+      if (as.size <= capacity - q.length) ref.update(_.enqueue(as.toList)).unit
+      else STM.retry
+    }
+
+  /**
+   * Views the next element in the queue without removing it, retrying if the
+   * queue is empty.
+   */
+  def peek: STM[Nothing, A] =
     ref.get.flatMap(
-      _.lastOption match {
+      _.headOption match {
         case Some(a) => STM.succeed(a)
         case None    => STM.retry
       }
     )
 
   /**
-   * Checks if the Queue is empty
+   * Views the next element in the queue without removing it, returning `None`
+   * if the queue is empty.
    */
-  def isEmpty: STM[Nothing, Boolean] = ref.get.map(_.isEmpty)
+  def peekOption: STM[Nothing, Option[A]] =
```

Comment:
This is useful, and also equivalent to, `peek.map(Some(_)) orElse STM.succeed(None)`.

Reply:
Uh oh, our code is too compositional! Better switch to actors! 😉